src/Covenant/Expr.hs

```diff
@@ -399,6 +436,15 @@ idOf e = ExprBuilder $ do
 nextId :: Id -> Id
 nextId (Id w) = Id $ w + 1
 
+withType :: ExprType -> Id -> TypedId
+withType = flip TypedId
+
+-- | Fail with type error.
+-- (2025/02/02) Farseen: I haven't figured out how to add error handling in the ExprBuilder monad.
+-- Till then, this will stand in for error handling.
+typeError :: forall a. String -> a
+typeError s = error s
```

Comment:
@kozross, this is temporary. I need to replace this with a proper error management solution.
I was thinking I'll insert an `Either String` in `ExprBuilder` using `mtl`. Is this good? Do you have any other ideas?

Reply:
`Either` is a good idea, `String` is absolutely not. You want a structured sum type specifying what the error is, along with relevant information. 

If you want type errors in `ExprBuilder`, you would need to add an `ExceptT` into the stack. However, be careful of how you order these: `ExceptT e (State s) a` and `StateT s (Except e) a` are quite different monads, and you need to make sure you get the semantics you want. 

You could also consider a different approach: when we 'compile' an `ExprBuilder` into a graph, we currently return a result in `Maybe`. Instead, you could return the result in `Either CompileError`, which might be a bit easier.